### PR TITLE
feat(api): inbox - the new get notifications endpoint

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?
In this PR:

1.  Implemented the new Inbox endpoint `GET /inbox/notifications`:
   - cursor-based pagination with `limit`, `after` (number or id) params
   - filtering by: tags, read, archived fields
2. The reusable `cursorPagination` function in the base repository.
3. The InboxNotification type that matches with DSL.
4. New fields on the `Message` schema: `archived, archivedAt, tags`.
5. E2E + unit tests included.

### Screenshots
Screenshot 2024-06-20 at 12 08 56
Screenshot 2024-06-20 at 12 09 21

### CODE REVIEW

File: apps/api/src/app/inbox/e2e/get-notifications.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  StepTypeEnum,
+  ChannelCTATypeEnum,
+  TemplateVariableTypeEnum,
+  ActorTypeEnum,
+  SystemAvatarIconEnum,
+  ChannelTypeEnum,
+} from '@novu/shared';
+import { mapToDto } from '../utils/notification-mapper';
+
+describe('Get Notifications - /inbox/notifications (GET)', async () => {
```
Comment (LetItRock) [inline apps/api/src/app/inbox/e2e/get-notifications.e2e.ts:14]: the e2e tests for the new endpoint

File: apps/api/src/app/inbox/usecases/get-notifications/get-notifications.command.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { IsMongoIdOrNumber } from '../../../shared/validators/mongo-id-or-number';
+
+export class GetNotificationsCommand extends EnvironmentWithSubscriber implements CursorPaginationParams {
+  @IsInt()
+  @Min(1)
+  @Max(100)
+  readonly limit: number;
+
+  @IsOptional()
+  @IsMongoIdOrNumber()
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/get-notifications/get-notifications.command.ts]: the new custom validator that checks if the value is Mongo ID or a number

File: apps/api/src/app/inbox/usecases/get-notifications/get-notifications.spec.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    subscriber: mockSubscriber,
+    actorSubscriber: mockSubscriber,
+    cta: {
+      type: ChannelCTATypeEnum.REDIRECT,
+      data: {},
+    },
+  },
+];
+
+describe('GetNotifications', () => {
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/get-notifications/get-notifications.spec.ts:33]: the unit tests

File: apps/api/src/app/inbox/usecases/get-notifications/get-notifications.usecase.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+import { ApiException } from '../../../shared/exceptions/api.exception';
+import { GetSubscriber } from '../../../subscribers/usecases/get-subscriber';
+import type { GetNotificationsResponseDto } from '../../dtos/get-notifications-response.dto';
+import { AnalyticsEventsEnum } from '../../utils';
+import { mapToDto } from '../../utils/notification-mapper';
+import type { GetNotificationsCommand } from './get-notifications.command';
+
+@Injectable()
+export class GetNotifications {
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/get-notifications/get-notifications.usecase.ts:13]: the usecase used in the controller

File: apps/api/src/app/inbox/utils/notification-mapper.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    channelType: channel,
+    tags,
+  };
+};
+
+export function mapToDto(notification: MessageEntity): InboxNotification;
+export function mapToDto(notification: MessageEntity[]): InboxNotification[];
+export function mapToDto(notification: MessageEntity | MessageEntity[]): InboxNotification | InboxNotification[] {
+  return Array.isArray(notification) ? notification.map((el) => mapSingleItem(el)) : mapSingleItem(notification);
+}
```
Comment (LetItRock) [inline apps/api/src/app/inbox/utils/notification-mapper.ts:73]: the helper function that is used in the usecase to map the `message` documents to the `notification` objects

File: apps/api/src/app/inbox/utils/types.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  subscriberId: string;
+};
+
+type Action = {
+  type: ButtonTypeEnum;
+  label: string;
+  url?: string;
+};
+
+export type InboxNotification = {
```
Comment (LetItRock) [inline apps/api/src/app/inbox/utils/types.ts:22]: the new Notification type that matches with latest DSL

File: apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts [hunk 7]
```diff
@@ -40,6 +43,14 @@ export class RunJob {
     let job = await this.jobRepository.findOne({ _id: command.jobId, _environmentId: command.environmentId });
     if (!job) throw new PlatformException(`Job with id ${command.jobId} not found`);
 
+    const template = await this.getNotificationTemplate({
+      _id: job._templateId,
+      environmentId: job._environmentId,
+    });
+    if (!template) {
+      throw new PlatformException(`Notification template ${job._templateId} is not found`);
+    }
```
Comment (LetItRock) [inline apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts:52]: Find the `template` when the job is running and propagate it's `tags` to the place where the `message` document is created

File: apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts [hunk 8]
```diff
@@ -121,6 +133,17 @@ export class RunJob {
...
+  @CachedEntity({
+    builder: (command: { _id: string; environmentId: string }) =>
+      buildNotificationTemplateKey({
+        _environmentId: command.environmentId,
+        _id: command._id,
+      }),
+  })
+  private async getNotificationTemplate({ _id, environmentId }: { _id: string; environmentId: string }) {
+    return await this.notificationTemplateRepository.findById(_id, environmentId);
+  }
```
Comment (LetItRock) [inline apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts:146]: find with caching

File: libs/dal/src/repositories/base-repository.ts [hunk 9]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+        isSortDesc,
+        paginateField,
+        after,
+        queryOrStatements: query?.$or,
+      });
+
+      findQueryBuilder = this.MongooseModel.find({
+        ...query,
+        $or: orStatements,
+      });
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts:193]: When the `after` property is `id` we would like to build the cursor-based query statements.
In MongoDB that statements usually looks like this:
```
{
   ...query,
   $or: [{
      createdAt: isSortDesc ? { $lt: afterCreatedAt } : { $gt: afterCreatedAt },
   }, {
      createdAt: { $eq: afterCreatedAt },
      _id: isSortDesc ? { $lt: after } : { $gt: after },
   }],
}
```

But this is the code of the base-repository, meaning you can pass your own query with an `$or` statement, so in that case in order for the cursor-based pagination to work, we need to merge these or statements with passed query or statements, so like this:
```
{
   ...query,
   $or: [{
      archived: { $exists: false },
      createdAt: isSortDesc ? { $lt: afterCreatedAt } : { $gt: afterCreatedAt },
   }, {
      archived: { $in: [true, false] },
      createdAt: isSortDesc ? { $lt: afterCreatedAt } : { $gt: afterCreatedAt },
   }, {
      archived: { $exists: false },
      createdAt: { $eq: afterCreatedAt },
      _id: isSortDesc ? { $lt: after } : { $gt: after },
   }, {
      archived: { $in: [true, false] },
      createdAt: { $eq: afterCreatedAt },
      _id: isSortDesc ? { $lt: after } : { $gt: after },
   }],
}
```

File: libs/dal/src/repositories/base-repository.ts [hunk 10]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
     }
   }
 
+  private async createCursorBasedOrStatement({
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts:113]: please read the comment from line 191 first

File: libs/dal/src/repositories/base-repository.ts [hunk 11]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+    }
+
+    let cursorOrStatements: FilterQuery<T_DBModel>[] = [];
+    let enchancedCursorOrStatements: FilterQuery<T_DBModel>[] = [];
+    if (paginateField && afterItem[paginateField]) {
+      const paginatedFieldValue = afterItem[paginateField];
+      cursorOrStatements = [
+        { [paginateField]: isSortDesc ? { $lt: paginatedFieldValue } : { $gt: paginatedFieldValue } } as any,
+        { [paginateField]: { $eq: paginatedFieldValue }, _id: isSortDesc ? { $lt: after } : { $gt: after } },
+      ];
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts:136]: the `or` statements based on the paginated field - used in the case when no custom `or` statements are passed in the `query`

File: libs/dal/src/repositories/base-repository.ts [hunk 12]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+      const firstStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[0],
+      }));
+      const secondStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[1],
+      }));
+      enchancedCursorOrStatements = [...firstStatement, ...secondStatement];
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts]: when custom `or` statements are passed in the `query` we need to merge it with cursor-based `or` statements

File: libs/dal/src/repositories/base-repository.ts [hunk 13]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+    } else {
+      cursorOrStatements = [{ _id: isSortDesc ? { $lt: after } : { $gt: after } }];
+      const firstStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[0],
+      }));
+      enchancedCursorOrStatements = [...firstStatement];
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts]: similar to above but for the case when we want to paginate just by the `_id` field

File: libs/dal/src/repositories/base-repository.ts [hunk 14]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+        queryOrStatements: query?.$or,
+      });
+
+      findQueryBuilder = this.MongooseModel.find({
+        ...query,
+        $or: orStatements,
+      });
+    }
+
+    findQueryBuilder.sort(sort).limit(limit + 1);
```
Comment (LetItRock) [inline libs/dal/src/repositories/base-repository.ts:196]: take a one more item from the `limit` - this allows us to understand if there are more items available for the next page

File: libs/dal/src/repositories/message/message.repository.ts [hunk 15]
```diff
@@ -96,12 +96,71 @@ export class MessageRepository extends BaseRepository<MessageDBModel, MessageEnt
...
+
+    if (typeof archived === 'boolean') {
+      if (!archived) {
+        query.$or = [{ archived: { $exists: false } }, { archived: false }];
+      } else {
+        query.archived = true;
+      }
+    } else {
+      query.$or = [{ archived: { $exists: false } }, { archived: { $in: [true, false] } }];
+    }
```
Comment (LetItRock) [inline libs/dal/src/repositories/message/message.repository.ts:148]: Handle the `archive` filtering, the thing is that it's a new field that is by default `undefined` for all the documents, which means the same as `archived: false`.

File: apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import type { InboxNotification } from '../utils/types';
+
+export class GetNotificationsResponseDto {
+  data: InboxNotification[];
+  hasMore: boolean;
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts:5]: @LetItRock I was under the impression we will also add this with a `nextPage` API url. If null, it means there is nothing more to load. [ref](https://docs.stripe.com/api/pagination/search)

File: apps/api/src/app/inbox/utils/notification-mapper.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    lastName: subscriber?.lastName,
+    avatar: subscriber?.avatar,
+    subscriberId: subscriber?.subscriberId ?? '',
+  };
+  const primaryCta = cta.action?.buttons?.find((button) => button.type === ButtonTypeEnum.PRIMARY);
+  const secondaryCta = cta.action?.buttons?.find((button) => button.type === ButtonTypeEnum.SECONDARY);
+
+  return {
+    id: _id,
+    // subject: subject,
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/utils/notification-mapper.ts:33]: Let's add a todo to support this next week as it directly affects the UI of the new Inbox.

File: apps/api/src/app/inbox/utils/notification-mapper.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    body: content as string,
+    to,
+    read,
+    archived,
+    createdAt,
+    readAt: lastReadDate,
+    archivedAt,
+    actor: actorSubscriber
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/utils/notification-mapper.ts:41]: PTAL at [this](https://www.notion.so/Notifications-DSL-dc7b2a8c7c6b49bf95e770e7c94d9faa?d=8e22b8a722c44f348a40d759a6df3d21&pvs=4#42c8642c1b2b4688979a98a8396db5e0).

File: apps/api/src/app/inbox/utils/notification-mapper.ts (hunk 5 above)
Comment (SokratisVidros) [reply apps/api/src/app/inbox/utils/notification-mapper.ts:73]: Nice. Let's add a comment explaining that as it goes against our current modelling (we will revise the discrepancies between messages and notification across the board soon).

File: apps/api/src/app/inbox/utils/types.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+  id: string;
+  subject?: string;
+  body: string;
+  to: Subscriber;
+  read?: boolean;
+  archived?: boolean;
+  createdAt: string;
+  readAt?: string;
+  archivedAt?: string;
+  actor?: Subscriber;
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/utils/types.ts:32]: Ditto - [ref](https://www.notion.so/Notifications-DSL-dc7b2a8c7c6b49bf95e770e7c94d9faa?d=8e22b8a722c44f348a40d759a6df3d21&pvs=4#42c8642c1b2b4688979a98a8396db5e0)

File: apps/api/src/app/shared/types.ts [hunk 20]
```diff
@@ -1 +1,6 @@
 export type Constructor<I> = new (...args: any[]) => I;
+
+export type CursorPaginationParams = {
+  limit: number;
+  after: number | string;
```
Comment (SokratisVidros) [inline apps/api/src/app/shared/types.ts]: I thought about this again, it's better if we split this into two parameters creating a combo of limit, offset, and limit, after.

File: libs/dal/src/repositories/base-repository.ts [hunk 21]
```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
...
+    after: string;
+    queryOrStatements?: object[];
+  }): Promise<FilterQuery<T_DBModel>[]> {
+    const afterItem = await this.MongooseModel.findOne({ _id: after });
+    if (!afterItem) {
+      throw new DalException('Invalid after id');
+    }
+
+    let cursorOrStatements: FilterQuery<T_DBModel>[] = [];
+    let enchancedCursorOrStatements: FilterQuery<T_DBModel>[] = [];
```
Comment (SokratisVidros) [inline libs/dal/src/repositories/base-repository.ts]: enhanced

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: apps/api/src/app/inbox/usecases/session/session.usecase.ts [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+
+    if (!environment) {
+      throw new ApiException('Please provide a valid application identifier');
+    }
+
+    const inAppIntegration = await this.selectIntegration.execute(
+      SelectIntegrationCommand.create({
+        environmentId: environment._id,
+        organizationId: environment._organizationId,
+        userId: command.subscriberId,
```
Comment (scopsy) [inline apps/api/src/app/inbox/usecases/session/session.usecase.ts]: @LetItRock we need to remove the userId from here, as it's causing issues with Cache imploding. A fix was already pushed to next for the existing session initialize endpoint

File: apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts (hunk 16 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts:5]: I've implemented it based on the API DX guide, but I can adjust it with this suggestion.
However, I'm not sure if it links well with the `after` property.

File: apps/api/src/app/inbox/utils/notification-mapper.ts (hunk 17 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/utils/notification-mapper.ts:33]: let's talk about that together

File: apps/api/src/app/inbox/utils/notification-mapper.ts (hunk 18 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/utils/notification-mapper.ts:41]: `actor` is also a subscriber from our system: https://docs.novu.co/additional-resources/glossary#actor

File: apps/api/src/app/shared/types.ts (hunk 20 above)
Comment (LetItRock) [reply apps/api/src/app/shared/types.ts]: which one would take the priority when both passed? `offset` or `after`?

File: apps/api/src/app/inbox/utils/types.ts (hunk 19 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/utils/types.ts:32]: https://github.com/novuhq/novu/pull/5792#discussion_r1651478406

File: apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts (hunk 16 above)
Comment (SokratisVidros) [reply apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts:5]: We can do it ala stripe. We can return the current `url` and a `next_url`.

File: apps/api/src/app/shared/types.ts (hunk 20 above)
Comment (SokratisVidros) [reply apps/api/src/app/shared/types.ts]: Throw a validation error (422) if both are passed.

File: apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts (hunk 16 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/dtos/get-notifications-response.dto.ts:5]: we decided to leave this for now

File: apps/api/src/app/inbox/utils/notification-mapper.ts [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+      ? {
+          id: actorSubscriber._id,
+          firstName: actorSubscriber.firstName,
+          lastName: actorSubscriber.lastName,
+          avatar: actorSubscriber.avatar,
+          subscriberId: actorSubscriber.subscriberId,
+        }
+      : undefined,
+    avatar: actor,
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/utils/notification-mapper.ts:50]: Maybe a better name for this param, as it's a bit confusing?

**Review by SokratisVidros — APPROVED** [review]

File: apps/api/src/app/inbox/utils/notification-mapper.ts (hunk 23 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/utils/notification-mapper.ts:50]: just the `icon`?

**Pull request merged**